### PR TITLE
WELD-1914 Add two interfaces denoting Weld-created constructs and all…

### PR DESCRIPTION
**Pull request by manovotn**
…owing to regain contextual instance and Bean metadata.

### CODE REVIEW

**Comment by manovotn** [comment]
Core PR draft -> https://github.com/weld/core/pull/1713

File: weld/src/main/java/org/jboss/weld/proxy/WeldConstruct.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jboss.weld.proxy;
+
+/**
+ * A marker interface denoting that the instance uses internal Weld contruct. This ranges from interceptor/decorator subclass to
```
Comment (mkouba) [inline weld/src/main/java/org/jboss/weld/proxy/WeldConstruct.java]: _the instance ~uses~ **is an** internal Weld contruct_
_This ranges from ~interceptor/decorator~ **intercepted/decorated** subclass_

File: weld/src/main/java/org/jboss/weld/proxy/WeldClientProxy.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jboss.weld.proxy;
+
+import javax.enterprise.inject.spi.Bean;
+
+/**
+ * An extension to WeldConstruct interface which denotes a client proxy object. All Weld client proxies implement this interface
```
Comment (mkouba) [inline weld/src/main/java/org/jboss/weld/proxy/WeldClientProxy.java]: Please use javadoc links, e.g. `{@link WeldConstruct}`

File: weld/src/main/java/org/jboss/weld/proxy/WeldClientProxy.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+    /**
+     * Returns Bean<?> metadata from this client proxy.
+     *
+     * @return Bean<?> for this proxy
+     */
+    Bean<?> getBean();
+
+    /**
+     * Returns underlying contextual instance of this client proxy object.
```
Comment (mkouba) [inline weld/src/main/java/org/jboss/weld/proxy/WeldClientProxy.java]: I would rather place this sentence in the `@return` tag...

File: weld/src/main/java/org/jboss/weld/proxy/WeldClientProxy.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+         *
+         * @return {@link Bean<?>} metadata for this proxy
+         */
+        Bean<?> getBean();
+
+        /**
+         * Retrieve the current contextual instance associated with the current context for this client proxy. Note that in some
+         * cases the contextual instance is still an instance of {@link WeldConstruct}.
+         *
+         * @return Returns the underlying contextual instance of this client proxy instance.
```
Comment (mkouba) [inline weld/src/main/java/org/jboss/weld/proxy/WeldClientProxy.java]: Superfluous _Returns_...

File: weld/src/main/java/org/jboss/weld/proxy/WeldConstruct.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jboss.weld.proxy;
+
+/**
+ * Marker interface denoting that the instance is an internal Weld contruct. This ranges from intercepted/decorated subclass to
```
Comment (mkouba) [inline weld/src/main/java/org/jboss/weld/proxy/WeldConstruct.java]: Maybe we should be more clear that this construct is used for contextual instances and contectual references only. E.g.  something like _"This ranges from contextual instances of intercepted/decorated beans to client proxies, that is contextual references for normal scoped beans."_

Comment (manovotn) [inline weld/src/main/java/org/jboss/weld/proxy/WeldConstruct.java]: Good point, changed to `This ranges from contextual instances of intercepted/decorated beans to contextual references for normal scoped beans (client proxies).`

**Pull request merged**